hooks/useClickOutside.ts

```diff
@@ -0,0 +1,33 @@
+import { useEffect } from 'react';
+import type { RefObject } from 'react';
+
+// eslint-disable-next-line no-unused-vars
+type TypeHandler = (event: MouseEvent | TouchEvent) => void;
+
+type UseClickOutsideProps = {
+  ref: RefObject<HTMLElement>;
+  handler: TypeHandler;
+};
+
+export const useClickOutside = (props: UseClickOutsideProps) => {
+  useEffect(() => {
+    const listener = (event: MouseEvent | TouchEvent) => {
+      if (
+        !props.ref.current ||
+        props.ref.current.contains(event.target as Node)
+      ) {
+        return;
+      }
+      props.handler(event);
+    };
+
+    document.addEventListener('mousedown', listener, true);
+    document.addEventListener('touchstart', listener, true);
```

Comment:
Why mousedown and touchstarts vs click? (Just out of curiosity)

Reply:
If I don't say any error it's allow to work with mouse and touch

Reply:
You still need to get my question. In some cases, both `mousedown` and `touchstart` will be fired together, but you only want one event.

You shouldn't create event listeners for both events, only one for the corresponding matching event that is appropriate for your device.

```tsx
const clickEvent = 'ontouchstart' in document ? 
  'ontouchstart' : 'onmousedown';

document.addEventListener(clickEvent, listener, { passive: true });

return () => document.removeEventListener(clickEvent, listener);
```

Reply:
I'm sure you probably just copied this example on StackOverflow because this code is 99% similar to StackOverflow, which is OK. Still, when copying things from StackOverflow, at least search and try to understand why certain things are done. And if they should be done like that.

Reply:
Your proposal work but with `in window`

Reply:
StackOverflow / GitHub repo = Doesn't matter 😅

The gist of my message is before simply pasting the code, think about why things are done the way they are and if anything should change. A good Engineer questions code, but also a sound Engineer knows when to listen.